### PR TITLE
[v6r12] Proposal to modify PilotStatusAgent to consider pilot's payload job

### PR DESCRIPTION
**Pull request by Kiyoshi-Hayasaka**
Hi,
We like to propose to modify PilotStatusAgent to consider pilot's payload job activity.
In the current code, even if pilot has running payload job, PilotStatusAgent force
the pilot deleted when the time limitation is exceeded. So, our modified code
to check the pilot's payload job activity before to make pilot deleted.

In addition, since PilotStatusAgent does not kill pilots actually,
"ghost" (or deleted) pilot makes its payload job execute later.  This makes some confusion.
So, we like to propose to introduce the code to kill pilots actually, whose status becomes deleted.

### CODE REVIEW

**Comment by chaen** [comment]
Hi,

Thanks for this PR. 

I cannot comment on whether this change is relevant or not.
However, as discussed in another PR, you should not instanciate a DB directly from the agent (JobDB). Please use the client associated.

**Comment by Kiyoshi-Hayasaka** [comment]
Hi,
I should use JobMonitoringHandler.
When I made the request, I could not find the corresponding handler
since there are several handlers to handle jobDB. Sorry!
I can modify it later. Thank you for pointing it out.

**Comment by fstagni** [comment]
To me it sounds a good idea. CC @atsareg

File: WorkloadManagementSystem/Agent/PilotStatusAgent.py [hunk 1]
```diff
@@ -505,3 +515,32 @@ def __addPilotsAccountingReport( self, pilotsData ):
...
+      if result['OK'] and result['Value'].has_key(i) and result['Value'][i].has_key('Status'):
+        ret = self.diracadmin.killPilot( str(i) )
+        if ret['OK']:
+          self.log.info("Sucsessfully deleted: %s (Status : %s)" % (i, result['Value'][i]['Status'] ) )
+        else:
+          self.log.error("Filed to delete %s : %s"  % ( i, ret['Message']))
+      else:
+        self.log.error("Failed to get info. of %s : %s" % ( i, str(result)))
+
+  def _CheckJobLastUpdateTime( self, joblist , StalledDays ):
```
Comment (atsareg) [inline WorkloadManagementSystem/Agent/PilotStatusAgent.py]: The DIRAC naming convention stipulates that method names must start with lower case character if even preceded by an underscore

**Comment by atsareg** [comment]
This is an interesting case, I mean very long living pilots in Running state. I guess you got it when running on non-grid sites where you do not have limitations in the task queue. Even if I agree in principle with the solution that you propose, may be in it is just simpler to increase the PilotStalledDays option value ( 3 by default ) to something more appropriate for your sites ?

**Comment by Kiyoshi-Hayasaka** [comment]
Thank you for the comments.
My point is that a disagreement between the number of running/waiting pilots on DIRAC and the actual one is so bad in the viewpoint from monitoring.
For example, when a site is so crowded and waiting time gets long, such pilots are "deleted" on DIRAC. On the other hand, the deleted waiting pilots are still alive on CE, later they accept payload jobs. On DIRAC, more running payload jobs than the number of the running pilots will be observed.

The solution is to kill the actual running pilot. That is realized by "_killPilots"
But, sometimes, actual running pilot is considered as stalled and DIRAC makes it "deleted" because this judgement is by only pilots' heart beat while this stops when its payload job is running. So, I like to introduce one more functionality, i.e., before making the judgement by pilots' heart beat, to check the running payload job's heart beat.

In the case where a pilot exceeds the limitation on a site, no problem happens because the pilot will be killed by the site actually.

As you pointed out, a site which allows longer CPU time 
(Since many of non-grid local sites have not-up-to-date cpus...)
has a tendency to make such ghost pilots.
I agree that this problem can be solved by making PilotStalledDays longer,
but, I do not like to solve this problem in such a way, since we may miss the chance to find the actually stalled pilots, in many cases, that indicates some hardware problem.

**Pull request merged**